packages/next/shared/lib/router/router.ts

```diff
@@ -110,6 +110,29 @@ function buildCancellationError() {
   })
 }
 
+function compareRouterStates(a: Router['state'], b: Router['state']) {
+  if (Object.keys(a).length !== Object.keys(b).length) return false
+
+  for (const key in a) {
+    if (key === 'query') {
+      if (Object.keys(a.query).length !== Object.keys(b.query).length) {
+        return false
+      }
+      for (const queryKey in a.query) {
+        if (a.query[queryKey] !== b.query[queryKey]) {
+          return false
+        }
+      }
+    } else if (
+      a[key as keyof Router['state']] !== b[key as keyof Router['state']]
+    ) {
+      return false
+    }
+  }
+
+  return true
+}
```

Comment:
```suggestion
function compareRouterStates(a: Router['state'], b: Router['state']) {
  const aKeys = Object.keys(a);
  const aKeysLen = aKeys.length;
  if (aKeysLen !== Object.keys(b).length) return false

  // It is most likely that shapes of parsed queries are different
  const aQueryKeys = Object.keys(a.query);
  const aQueryKeysLen = aKeys.length;
  if (aQueryKeysLen !== Object.keys(b.query).length) return false
  for (let j = aQueryKeysLen; j-- !== 0;) {
    if (!Object.prototype.hasOwnProperty.call(b.query, aQueryKeys[j])) return false
  }
  // Only compare the value when shapes are the same
  for (let j = aQueryKeysLen; j-- !== 0;) {
    const queryKey = aQueryKeys[j]
    if (a.query[queryKey] !== b.query[queryKey]) return false;
  }

  for (let i = aKeysLen; i-- !== 0;) {  
    const key = aKeys[i];
    if (key === 'query') continue;

    // It is less likely that shapes of router states are different
    // So compare the shape and the value together
    if (
      !Object.prototype.hasOwnProperty.call(b, key)
      || a[key as keyof Router['state']] !== b[key as keyof Router['state']]
    ) {
      return false;
    }
  }

  return true
}
```

A more performant version of `compareRouterStates`.

Reply:
Thanks! Would you mind explaining why is it more performant? Because it checks `query` before other properties?

Reply:
> Thanks! Would you mind explaining why is it more performant? Because it checks `query` before other properties?

It contains serveral optimizations:

- Use `for .. in` on an object will iterate over object's prototype. Use `for .. in` on `Object.keys` fixes that.
- `Route['state'].query` comes from `querystring.parse` where the shape is likely to change. The code checks the shape of the object before checking the value.
- The shape of `Route['state']` is not likely to change, so it is not necessary to check the shape before value.
- The implementation caches the result of `Object.keys` (and its length)

Reply:
Thanks for the reply! Could be wrong but here's what I think:
- I think the state is created as literal object, hence I didn't add the `hasOwnProperty` check;
- "The code checks the shape of the object before checking the value. " — Makes sense, I think those two loops from your code can be merged too?
- Regarding state and query, I'm not sure which one is more likely to change. In many cases that a route state change is caused by pushing a new route, where the `.route` value updates;
- "The implementation also caches / re-use the result of Object.keys" — I don't think this is a problem in the code from this PR?

Reply:
> * I think the state is created as a literal object, hence I didn't add the `hasOwnProperty` check;

Since `Object.keys` is being used already, IMHO it make sense to re-use it (and also prevent potential prototype literation)

> * "The code checks the shape of the object before checking the value. " — Makes sense, I think those two loops from your code can be merged too?

I haven't run a benchmark yet, but yes.

> * Regarding state and query, I'm not sure which one is more likely to change. In many cases that a route state change is caused by pushing a new route, where the `.route` value updates;

My original idea was that we are not re-creating a fast-shallow-equal library, the `compareRouterStates` function should be optimized specifically for route state.

> * "The implementation also caches / re-use the result of Object.keys" — I don't think this is a problem in the code from this PR?

It is just a micro-optimization in my code.

Reply:
Let me polish it a bit, thanks! :)

Reply:
One thing to point out here though, if two projects have the same number of own properties (`Object.keys(a).length === Object.keys(b).length`), and all values from one object are equal to the value from the other, these two objects are equal. We don't need to check their shapes.

Reply:
> One thing to point out here though, if two projects have the same number of own properties (`Object.keys(a).length === Object.keys(b).length`), and all values from one object are equal to the value from the other, these two objects are equal. We don't need to check their shapes.

Make sense. But since both `fast-equals`, `fast-deep-equals` and `fast-shallow-equals` run `Object.prototype.hasOwnproperty` before comparing the value, I guess there is a reason of they doing it.

We could skip the shape check for the entire router state (its shape is not likely to change) though.

Reply:
Yeah, the exception will be something like this:

```js
const a = { foo: 1 }
const b = Object.create({ foo: 1 })
b.bar = 1
```

Although it should not happen in our specific context, but totally understand your concern here. Since router state  comparison isn’t a heavy operation, I’ll apply the hasOwnProperty check here.

Thanks again!